src/interpreter/runtime/values/number.js

```diff
@@ -1,167 +1,87 @@
-// @flow
 import { RuntimeError } from "../../../errors";
 
-export class BaseNumber implements NumberInterface<BaseNumber> {
+// @flow
+export class Float<U> implements FloatingPointValue<Float<U>, U> {
   _value: number;
 
   constructor(value: number) {
     this._value = value;
   }
 
-  add(operand: BaseNumber): BaseNumber {
-    return new BaseNumber(this._value + operand._value);
+  add(operand: Float<U>): Float<U> {
+    return new this.constructor(this._value + operand._value);
```

Comment:
That an interesting hack here :smile:

Reply:
Yeah, polymorphism / inheritance isn't JavaScript's strong point!